### DISCUSSION TITLE
What is the reason for the slightly different result with Stab2D?

### DISCUSSION THREAD

**Discussion by Martin15135215** [adtzlr/trusspy#60] [Q&A] [answered]
This is the truss example:

![image](https://github.com/adtzlr/trusspy/assets/110968398/1bba1994-3609-4640-a686-ae3c5674cf2e)

This is the TrussPy code:

```python
import trusspy as tp

M = tp.Model(log=0)

# create nodes
with M.Nodes as MN:
    MN.add_node(1, (0, 0, 0))  # mm
    MN.add_node(2, (0, 0, 3000))  # mm
    MN.add_node(3, (-4000, 0, 3000))  # mm
    MN.add_node(4, (-4000, 0, 6000))  # mm

# create element
with M.Elements as ME:
    ME.add_element(1, [2, 1])
    ME.add_element(2, [2, 3])
    ME.add_element(3, [2, 4])
    ME.assign_geometry("all", [1000])  # Area mm²
    ME.assign_material("all", [1000])  # E-Modulus N/mm²


# create displacement (U) boundary conditions
with M.Boundaries as MB:
    MB.add_bound_U(1, (0, 0, 0))  # support
    MB.add_bound_U(2, (1, 0, 1))  # smooth pin
    MB.add_bound_U(3, (0, 0, 0))  # support
    MB.add_bound_U(4, (0, 0, 0))  # support


# create external forces
with M.ExtForces as MF:
    MF.add_force(2, (100000, 0, -100000))  # N

# TrussPy is a nonlinear truss analysis - it incrementally searches for the equilbrium
# by scaling given external forces. If you are not interested in the deformation path,
# set the number of increments ``incs`` to one, the allowed increase in displacements to
# infinity or a high value ``du`` and the increase of the load-proportionality-factor
#  ``dlpf`` to one.
M.Settings.incs = 1  # evaluate only one increment
M.Settings.du = 1e8  # turn off max. incremental displacement
M.Settings.dlpf = 1  # apply the load proportionality factor in one increment

M.build()

fig, ax = M.plot_model(force_scale=0.01, view="xz")

M.run()

# results of last solution
res = M.Results.R[-1]  

# verify the applied load-proportionality-factor of the result
# assert res.lpf == 1.0  

# (element) truss forces in N
# see https://trusspy.readthedocs.io/en/latest/theory/truss.html#kinetics
res.element_force

# (nodal) displacements in mm
# see e.g. https://trusspy.readthedocs.io/en/latest/theory/equilibrium.html
res.U

# (nodal) fixing forces in N
# see https://trusspy.readthedocs.io/en/latest/theory/assembly.html
res.r
```

Those are the TrussPy results:

![image](https://github.com/adtzlr/trusspy/assets/110968398/6b98b6d5-99fe-485d-8c90-9b7d66e761f9)

Now I build the same truss in [Stab2D](https://www.isd.uni-hannover.de/de/studium/software/). This software can be freely downloaded for educational purpose etc.

![image](https://github.com/adtzlr/trusspy/assets/110968398/c1c73382-2d77-40ea-bf5f-93d75319b036)

![image](https://github.com/adtzlr/trusspy/assets/110968398/331f36be-2cb2-43fa-9774-cbc64cae7cff)

```
Stab2D - Ausdruck der Ergebnisse

Dieses Programm ist lizensiert für: Dies ist eine Demo-Version.
Adresse                           : Tel. +49 511 762 3867   Fax +49 511 762 2236
8:55:17 PM  11/10/2023

Ergebnisse: I.Ordnung


---------------------------------- Knotenverschiebung ------------------------------
Knoten      wx         wz     Rotation
------------------------------------------------------------------------------------
    1     -0.000     -0.000    0.08365
    2      214.8      195.8    0.04751
    3     -0.000     -0.000    0.04968
    4     -0.000     -0.000   -0.01542

------------------- Verschiebung des Stabes   1 in 1/5   Punkten -------------------
xi-Position      wx         wz
------------------------------------------------------------------------------------
    0.2000      182.3      156.7
    0.4000      142.8      117.5
    0.6000      98.07      78.33
    0.8000      49.90      39.17

------------------- Verschiebung des Stabes   2 in 1/5   Punkten -------------------
xi-Position      wx         wz
------------------------------------------------------------------------------------
    0.2000      171.9      157.5
    0.4000      128.9      118.6
    0.6000      85.93      79.30
    0.8000      42.96      39.72

------------------- Verschiebung des Stabes   3 in 1/5   Punkten -------------------
xi-Position      wx         wz
------------------------------------------------------------------------------------
    0.2000      190.0      132.5
    0.4000      153.1      85.28
    0.6000      107.1      50.14
    0.8000      55.05      23.05


-----------------------Zugehörige Schnittkräfte -----------------------------------
Stab Nr.1
   Schnittkraft   x-Pos           N           Q           M
Max       N       0.000  -6.528e+04   8.030e-06    -0.02409
Min       N       0.000  -6.528e+04   8.030e-06    -0.02409

Max       Q       0.000  -6.528e+04   8.030e-06    -0.02409
Min       Q       0.000  -6.528e+04   8.030e-06    -0.02409

Max       M       1.000  -6.528e+04   8.030e-06  -3.082e-17
Min       M       0.000  -6.528e+04   8.030e-06    -0.02409

Stab Nr.2
   Schnittkraft   x-Pos           N           Q           M
Max       N       0.000   5.370e+04   2.709e-07   -0.001084
Min       N       0.000   5.370e+04   2.709e-07   -0.001084

Max       Q       0.000   5.370e+04   2.709e-07   -0.001084
Min       Q       0.000   5.370e+04   2.709e-07   -0.001084

Max       M       1.000   5.370e+04   2.709e-07       0.000
Min       M       0.000   5.370e+04   2.709e-07   -0.001084

Stab Nr.3
   Schnittkraft   x-Pos           N           Q           M
Max       N       0.000   5.787e+04  -5.035e-06     0.02517
Min       N       0.000   5.787e+04  -5.035e-06     0.02517

Max       Q       0.000   5.787e+04  -5.035e-06     0.02517
Min       Q       0.000   5.787e+04  -5.035e-06     0.02517

Max       M       0.000   5.787e+04  -5.035e-06     0.02517
Min       M       1.000   5.787e+04  -5.035e-06   9.012e-18


------------------ Schnittkräfte des Stabes   1 in 1/5   Punkten -------------------
 Xi-Pos          N          Q          M  
  0.000 -6.528e+04  8.030e-06   -0.02409
 0.2000 -6.528e+04  8.030e-06   -0.01927
 0.4000 -6.528e+04  8.030e-06   -0.01445
 0.6000 -6.528e+04  8.030e-06  -0.009637
 0.8000 -6.528e+04  8.030e-06  -0.004818
  1.000 -6.528e+04  8.030e-06 -3.082e-17

------------------ Schnittkräfte des Stabes   2 in 1/5   Punkten -------------------
 Xi-Pos          N          Q          M  
  0.000  5.370e+04  2.709e-07  -0.001084
 0.2000  5.370e+04  2.709e-07 -0.0008668
 0.4000  5.370e+04  2.709e-07 -0.0006501
 0.6000  5.370e+04  2.709e-07 -0.0004334
 0.8000  5.370e+04  2.709e-07 -0.0002167
  1.000  5.370e+04  2.709e-07      0.000

------------------ Schnittkräfte des Stabes   3 in 1/5   Punkten -------------------
 Xi-Pos          N          Q          M  
  0.000  5.787e+04 -5.035e-06    0.02517
 0.2000  5.787e+04 -5.035e-06    0.02014
 0.4000  5.787e+04 -5.035e-06    0.01510
 0.6000  5.787e+04 -5.035e-06    0.01007
 0.8000  5.787e+04 -5.035e-06   0.005035
  1.000  5.787e+04 -5.035e-06  9.012e-18

------------------------------------- Auflagerkräfte -------------------------------
Knoten       x          z     Rotation
------------------------------------------------------------------------------------
    1 -8.030e-06 -6.528e+04  3.082e-17
    3 -5.370e+04 -2.709e-07      0.000
    4 -4.630e+04 -3.472e+04 -9.012e-18

---------------------------------- Gelenkverschiebung ------------------------------
Gelenk linke Seite  rechte Seite Gesamt
------------------------------------------------------------------------------------

```

For example, the x-axis displacement is 10 mm off

**Comment by adtzlr** [comment]
Does this software give you the linearized solution? The code I provided for TrussPy is the nonlinear result, even for one increment, nonlinear geometric exact kinematics are used. The result is obtained by a Newton Rhapson iterative method. The linearized solution could be slightly different (depends on the model). That could be one reason.

We could enforce the linear result in TrussPy by setting the tolerance of the equilibrium equations to infinity, which will stop the Newton method after one iteration. I'll have a look later on how to set that parameter.

**Reply by Martin15135215** [reply]
I don't know what Stab2D did, but I have an Excel with this task from last semester. The stiffness method was used in Excel. The Excel was created by my statics lecturer to show us the stiffness method. And I get the same values in Stab2D as in Excel.

**Reply by adtzlr** [reply]
A free web application (https://valdivia.staff.jade-hs.de/fachwerk_en.html) gives also the same results as Stab2D.

![image](https://github.com/adtzlr/trusspy/assets/5793153/2995b3cf-dd99-4988-9a30-8b1f16d90c98)

I'll have to dive into the code again 🕵🏻

**Comment by adtzlr** [comment, marked as answer]
Okay, I figured out that the differences in the results are due to effects of nonlinearity. Equal results with Stab2D are obtained only by setting the load-proportionality factor to a very small value in TrussPy.

Due to the nonlinear path-following algorithm, extended equilibrium equations are used for the numeric continuation. However, there is no direct access implemented to the linearized solution other than applying very small loads which lead to small displacements and (nearly) no effects of nonlinearity.

With `M.Settings.dlpf = 1e-5` you'll get the displacements

```IPython
In[1]: np.round(res.U / res.lpf, 1)
Out[2]: 
array([[   0. ,    0. ,    0. ],
       [ 214.8,    0. , -195.8],
       [   0. ,    0. ,    0. ],
       [   0. ,    0. ,    0. ]])
```

as well as the fixing forces

```IPython
In[3]: np.round(res.r / res.lpf, 0)
Out[4]: 
array([[      0.,       0.,   65278.],
       [ 100000.,       0., -100000.],
       [ -53704.,       0.,       0.],
       [ -46296.,       0.,   34722.]])
```

and the element forces which are all consistent with Stab2D.

```IPython
In[5]: np.round(res.element_force / res.lpf, 0)
Out[6]: 
array([[-65278.],
       [ 53704.],
       [ 57870.]])
```

FYI: if you are interested in the stiffness matrix for the active degrees of freedom, it is located here

```IPython
In[7]: res.Kred
Out[8]: 
array([[377.99981341, -95.99983127],
       [-95.99983127, 405.33355233]])
```

from which the linear solution may be found manually by

```IPython
In[9]: np.linalg.solve(res.Kred, [100000, -100000])
Out[10]: array([ 214.81502508, -195.83326715])
```

Hope that helps!

**Reply by Martin15135215** [reply]
Thank you for your assistance!

1. Question:

Regarding `M.Settings.dlpf = 1e-5` it is the same value for any consistent unit input (kN,m or N,mm or MN,cm etc.) ?

2. Question:

The manual way: how can I get from the small displacement array to the (element) truss forces and (nodal) fixing forces and (nodal) displacements?

**Reply by adtzlr** [reply]
1. It does not depend on the units but on the applied loads. It must scale down the loads to result in small strains. Too small values could lead to numerical problems (round off errors).

2. TrussPy has no manual functions for that. For the theory, please have a look at the docs.

**Comment by Martin15135215** [comment]
> The code I provided for TrussPy is the nonlinear result, even for one increment, nonlinear geometric exact kinematics are used. The result is obtained by a Newton Rhapson iterative method. The linearized solution could be slightly different (depends on the model).

I have a question regarding this: Does TrussPy calculate all trusses in the first order theory or second order theory or third order theory or the full nonlinear theory? I don't know which of the theories you mean with “nonlinear result”.

Below are the definitions according to IFC.

> This enumeration is used to distinguish between different types of structural analysis methods, including first order theory, second order theory (small deformations), third order theory (large deformations) and the full nonlinear theory (geometric nonlinearity together with other nonlinearities such as plasticity).

Source: https://github.com/buildingSMART/IFC4.3.x-development/blob/master//docs/schemas/domain/IfcStructuralAnalysisDomain/Types/IfcAnalysisTheoryTypeEnum.md
